### ISSUE TITLE
[LLC]  Make `RequestOptions` required for operations without a request body.

### ISSUE THREAD

**Issue by chamons** [Azure/autorest.csharp#1332]

**Comment by annelo-msft** [comment]
Problem is that a Get method, e.g. could be called without parameters => ambiguous call site.
We need this to prevent ambiguities

=> We need to confirm the API design here.  Is this still a problem, given the grow-up story?
=> To know whether we need to do this for RC1.0 or not, we need to finalize the models story

**Comment by AlexanderSher** [comment]
Why do we need required `RequestOptions` in this case: https://github.com/Azure/autorest.testserver/blob/dc0fde45b200836099f79079427b71e5f9c81afd/swagger/httpInfrastructure.json#L78

**Comment by annelo-msft** [comment]
@ellismg

**Comment by ellismg** [comment]
The goal was to "leave space" in the method group for an HLC method that layered on models.  For example, consider the following case, an operation that returns a pet given an name (the name ends up being a path parameter in the service e.g. `<endpoint>/pets/<petName>`).  We would generate client with a method like this:

```csharp
public class PetsClient {
    public Response GetPet(string petName, RequestOptions requestOptions = default);
}
```

The thought was this design would make it difficult to add an overload that exposed a model type:


```csharp
public class PetsClient {
    public Response GetPet(string petName, RequestOptions requestOptions = default);

    /* New in V2, but now any that elided the requestOptions parameter, i. e. GetPet("Snoopy") are now ambiguous */
    public Response<Pet> GetPet(string petName, CancellationToken cancellationToken = default);
}
```

For operations which take request body, this isn't a problem because there will be a `RequestContent` parameter in the method signature in the protocol method which would be replaced with a model type for the request body when we added an overload that also returned a model.

However, we seem to be moving to a models story where maybe we don't add these overloads and instead expect customers to call the protocol method as before and then cast the response to a `Pet`.  If that's the long term plan we are committed to, we can just not do this work.  It would be good to confirm with @annelo-msft and @KrzysztofCwalina about the plan.

Note that even with the above design you could still run into source level ambiguities in cases where someone explicitly passed `default` as the second parameter.  We were ok with these source level breaking changes as we assumed such patterns would be uncommon in practice and the BCL faces similar issues as they evolve their APIs.

Note - if we end up not doing this and end up wanting to add an method that returns a model, we discussed using the `Value` suffix, e.g:

```csharp
public class PetsClient {
    public Response GetPet(string petName, RequestOptions requestOptions = default);
 
    /* New in V2 */
    public Response<Pet> GetPetValue(string petName, CancellationToken cancellationToken = default);
}
```

**Comment by annelo-msft** [comment]
When last @KrzysztofCwalina and I spoke about this, we had said we would not have a `CancellationToken` in the grow-up method (I believe we don't call these HLC methods as well anymore, as we're not shipping HLCs alongside LLCs, rather an LLC + "grow-up or helper methods").  This had been discussed in [this PR](https://github.com/ellismg/design-notes/pull/1#discussion_r695124458).

However, I know that ideas have been evolving quickly on this -- @KrzysztofCwalina, could you share where your latest thinking is on `CancellationToken` ?

**Comment by ellismg** [comment]
I think this is required, even if we decide we wouldn't have cancellation tokens, in fact the problem is worse because instead of a source ambiguity you end up just calling the newer overload, which returns a model:

```
public class PetsClient {
    public Response GetPet(string petName, RequestOptions requestOptions = default);

    /* New in V2, but now any that elided the requestOptions parameter, i. e. GetPet("Snoopy") now calls this method! */
    public Response<Pet> GetPet(string petName);
}
```

I think that would be very surprising to folks.  Folks who explicitly set the return type, i.e.:

```
Response petResponse = client.GetPet("Snoopy");
```

Would at least have the compiler complain (since Response is not a super type of Response<T>) but anyone who had used `var` is likely going to be very confused as to what's going on.

You can play around with [this example on SharpLab](https://sharplab.io/#v2:D4AQTAjAsAUCDMACciBKBTAjgV3QZwBcB5ABwIEsB7AOz0QG9EBfWBZMNfEmvdB5JOWoFEAZQIBDAtjwBhSgBM+jAOboRAXgB8iMAAY9AbmbNWSFBjzda6ADwAVHYxZxzHAArr+Ls+0SeCOQAbcnRhBlhEKLYAN3IAJ2kJIM4rHj4AcXUAgAoQCD1EEnUAOQkAW3QAGk4cfGIyKlpEeKxcQlIKHkQNRCUAMwlsIIIASh6damGgw0iogHp52ISklMtrXlsAnSyCXPzC4oIyyvHtRCmgmdgfVz97eoiYKOiwDno5l5eYiXjEAGMQmFNBd0AB3fzqYKhYQ5UazZ5fb6/FpcdI9AFA4QAOl2uQARKJqJRKCQAJ74+GfJH5ACcOUJkmkdHIeAAXIh8YgANSotI2bHiKQyeRKKmIqIuJhAA===). and see what happens when you uncomment the `GetPet` method which returns a model type.

**Comment by KrzysztofCwalina** [comment]
I am not sure what's worse: the source break, or having to pass RequestOptions to call calls on MVP (LLC without helpers).

**Comment by KrzysztofCwalina** [comment]
Also, this might be a bit too hacky, but the following works:

```csharp
var client = new Client();
Response r = client.Bar(); // this will call the parameterless method

public class OutModel
{
    public static implicit operator OutModel(Response response) => default;
    public static implicit operator Response(OutModel outModel) => default;
}
public class Client
{
    public Response Bar(RequestOptions options = default) => default;

    public OutModel Bar() => default;
}
```

**Comment by ellismg** [comment]
Yes - I do see how the above could side step the source issues.  If we are okay with the source breaking change for the `var` case when we introduce the new overload, that feels like something that could be workable.

One possible concern is that it would be harder to leverage `OutModel` as an input, since in that case there would be no `Response` to return.  Maybe we have the implicit cast operator return `null` in that case, things would hang together.  And maybe it makes sense given how our model story is shaping up.

Regarding the having to pass RequestOptions - I get that it's a bit of a wart.  If we made this behavior customizable at an operation level, we could pick a default of "make RequestOptions be a required parameter" but allow using an extension on the operation like  `x-ms-allow-default-request-options` (or some better name) which would let you emit a method where `RequestOptions` has a default value.

If long term we decided we never wanted to add models, we could add a helper with a default request options that just called into the version where it was required?

**Comment by KrzysztofCwalina** [comment]
> If long term we decided we never wanted to add models

What if we make a mistake? i.e. add a parameterless helper and then realize that we do need to add models. 

Also, it means LLC MVP is not as good as it could be, which makes people complain about LLC more. The complaint does not even go away when OOB models are available, as users have to keep passing RO.

**Comment by KrzysztofCwalina** [comment]
> if we are okay with the source breaking change for the var case when we introduce the new overload, that feels like something that could be workable.

Yeah, this is a good point. I am on the fence. 

BTw, a different hack I just came up with is to inherit Models from Response :-)

**Comment by ellismg** [comment]
> What if we make a mistake? i.e. add a parameterless helper and then realize that we do need to add models.

Then we have to pick a worse name for the model case.  I think a reasonable convention is to just use `Value`, so `GetPetValue`.  This isn't a hammer we'd want to use often and hopefully we wouldn't make the wrong decision multiple times, but I do think we have an escape hatch.  At that point we could even just consider doing things the "right way" using the `GetPetValue` name and then just hide the existing `GetPet` method group as much as we could (retaining it for binary compatibly).

**Comment by KrzysztofCwalina** [comment]
So I think the tradeoff looks like the following:

Option 1 (RequestOptions required):
1. (+) Models work great for both inputs and outputs
2. (-) Worst UX of MVP
3. (-) Possibly inconsistent method naming as the APIs evolve (some helpers are not overloads but rather suffixed with "Value")

Option 2 (RequestOptions optional)
1. (+) Consistent, good UX for MVP and good UX as the APIs evolve.
2. (-) Input-output models might fail to cast to Response.
3. (-) might be too clever design that causes unforeseen problems in the future.
4. (-) models have to carry Response, which makes them more complex, heavier, and cannot be used in the middle of a response object graph.

**Comment by annelo-msft** [comment]
Per design discussion, we will:

1. Make `RequestOptions` a required parameter for operations that return a payload in the response body.

This is because they would return `Response<T>` in an HLC, and we want to leave room for a corresponding helper method to be added as part of the grow-up story at some point in the future.

2. Make `RequestOptions` optional for any operation that does not return a payload in the response body.

It is rare that a REST API evolves to return a response payload where it did not return one before, so we expect we will never want to add a convenience helper that returns a model.

3. For `DELETE` operations, always make `RequestOptions` optional.  

This is because returning the deleted resource makes the request non-retriable.  We change this in HLC APIs in the customization layer; this will now bake that in to CodeGen.

4. We will provide the ability for the LLC developers to make `RequestOptions` optional on a method, based on input from the arch board in an [Arch Board Light Review](https://github.com/Azure/azure-sdk/issues/2734).

In C#, this will be accomplished by writing a partial class that changes the method definition.

**Comment by annelo-msft** [comment]
@AlexanderSher, can you confirm this is the design we want to go with before we assign it?

**Comment by AlexanderSher** [comment]
Yes, except I would like to have a flag in config that would make all `RequestOptions` optional. Making `RequestOptions` optional on a method is a great feature, but it requires some good amount of work, while flag is just several lines of code.

**Comment by annelo-msft** [comment]
> Yes, except I would like to have a flag in config that would make all `RequestOptions` optional. Making `RequestOptions` optional on a method is a great feature, but it requires some good amount of work, while flag is just several lines of code.

Yes, if arch board approves of making them all optional, I think it's great to be able to simplify the implementation.  I do believe we want to provide the option to do it on a per method basis as well, though.
